[RN] Fix keeping pinned participant across conferences

saghul opened the pull request.

lyubomir commented on react/features/base/participants/middleware.js:
```diff
@@ -32,9 +32,11 @@ MiddlewareRegistry.register(store => next => action => {
-    case CONFERENCE_LEFT:
+    case CONFERENCE_LEFT: {
         store.dispatch(localParticipantIdChanged(LOCAL_PARTICIPANT_DEFAULT_ID));
+        store.dispatch(pinParticipant(null));
```
I'm afraid I don't understand the offered solution. I think clearing the pinned redux state will be enough.

saghul replied:
I'll take another look.

saghul commented:
@lyubomir I changed the approach to act on the reducer instead of dispatching the extra action on the middleware. PTAL.

lyubomir commented on react/features/base/participants/reducer.js:
```diff
@@ -159,6 +164,9 @@ function _participant(state: Object = {}, action) {
     switch (action.type) {
+    case CONFERENCE_LEFT:
+        return state.map(p => _participant(p, action));
+
```
🤔

* Why `return state.map(p => _participant(p, action));` and the empty line after it? Isn't the `case` exactly the same as the 4 `case`s right after it? I think it's an unnecessary duplication which is in violation of our coding style.

* Why `CONFERENCE_LEFT` and why it only? I don't think we should be going on with `CONFERENCE_LEFT` anymore. We've talked how we often have to handle`CONFERENCE_LEFT` and `CONFERENCE_FAILED`. I think we should look into introducing actions for more general, catch-all events. For example, events similar to the Web's such as `beforeunload` (that'd be something around `SET_LOCATION_URL` I presume).

saghul commented:
> Why return state.map(p => _participant(p, action)); and the empty line after it? Isn't the case exactly the same as the 4 cases right after it? I think it's an unnecessary duplication which is in violation of our coding style.

My mistake, fixed it now. Thanks for pointing it out!

> Why CONFERENCE_LEFT and why it only? I don't think we should be going on with CONFERENCE_LEFT anymore. We've talked how we often have to handleCONFERENCE_LEFT and CONFERENCE_FAILED.

I now switched to CONFERENCE_WILL_JOIN, so we "cleanup" before joining. This avoids the left / failed situation. 

> I think we should look into introducing actions for more general, catch-all events. For example, events similar to the Web's such as beforeunload (that'd be something around SET_LOCATION_URL I presume).

100% agree there. That however is a much larger change which I'd rather tackle in a different PR if that's ok, since it will probably involve changing a lot more code.

Thanks for the review @lyubomir. I think it should be ready now, can you PTAL?

lyubomir commented on react/features/base/participants/reducer.js:
```diff
@@ -53,6 +54,10 @@ const PARTICIPANT_PROPS_TO_OMIT_WHEN_UPDATE
  */
 function _participant(state: Object = {}, action) {
     switch (action.type) {
+    case CONFERENCE_WILL_JOIN:
```
I'm afraid I don't find `CONFERENCE_WILL_JOIN` to be an acceptable action type though because it may very well arrive after I've been able to tap on a thumbnail.

saghul replied:
I don't think that's possible.

saghul replied:
I'll look further for an alternative.

lyubomir replied:
You mean it's not possible because we no longer as of yesterday evening render the filmstrip before a second participant joins? (If it's that, then I don't approve of such a dependency.) Or something else?

saghul commented:
wait

saghul commented:
Superseded by https://github.com/jitsi/jitsi-meet/pull/3794
